fish-rust/src/builtins/ulimit.rs

```diff
@@ -1,16 +1,489 @@
+use std::{cmp::Ordering, mem::MaybeUninit};
+
+use libc::{c_uint, rlim_t, rlimit, RLIM_INFINITY};
+use once_cell::sync::Lazy;
+
+use crate::{compat::*, fallback::*};
+
 use super::prelude::*;
 
+/// Safe wrapper around getrlimit
+fn getrlimit(resource: c_uint) -> Result<rlimit, i32> {
+    // TODO: better handle errno when using this function
+
+    use libc::getrlimit;
+    let mut rl: MaybeUninit<rlimit> = MaybeUninit::uninit();
+
+    // Only linux has resource type as __rlimit_resource_t = c_uint. Others use c_int
+    #[cfg(not(target_os = "linux"))]
+    let resource = resource as c_int;
+
+    let result = unsafe { getrlimit(resource, rl.as_mut_ptr()) };
+
+    if result == 0 {
+        Ok(unsafe { rl.assume_init() })
+    } else {
+        Err(result)
+    }
+}
+
+/// Safe wrapper around setrlimit
+fn setrlimit(resource: c_uint, rl: &rlimit) -> Result<(), i32> {
+    use libc::setrlimit;
+
+    // Only linux has resource type as __rlimit_resource_t = c_uint. Others use c_int
+    #[cfg(not(target_os = "linux"))]
+    let resource = resource as c_int;
+
+    let result = unsafe { setrlimit(resource, rl as *const rlimit) };
+
+    if result == 0 {
+        Ok(())
+    } else {
+        Err(result)
+    }
+}
```

Comment:
For these 2 functions, we can use the safe interfaces exposed by the [Nix](https://docs.rs/nix/latest/nix/sys/resource/fn.getrlimit.html) crate, so that we don't need to bother with the unsafe libc stuff

Reply:
> Just implemented this but still had to use unsafe to convert i32 into the Resource enum unfortunately

Yeah, this is unfortunate, maybe we should provide a `try_from(i32)` function here

Reply:
Don't know how feasible it is since the values from different RLIMITs can change across platforms. We'd be reimplementing nix's or libc's code with a try_from implementation. The values come directly from C and are either defined or -1 but when the transmutation happens it's already a uint, so it's safe to do so.

Reply:
> We'd be reimplementing nix's or libc's code with a try_from implementation

Yeah, implementing it in fish does not make sense, I was thinking about doing it at Nix. Just realized that the macro used in Nix already has this functionality, [but it requests you to specify a `repr`](https://github.com/nix-rust/nix/blob/39ad47bc80af68fb9ad67d37ab699684bea0e2a9/src/macros.rs#L316), which is not suitable for things that are different across OSes,  I am thinking about adding it using third-party macros like [`num_enum`](https://crates.io/crates/num_enum).

FYI, `Resource` is `u32` on Linux(gnu) and Linux(uclibc), on musl and macOS, it has repr `i32`:

```rs
    #[cfg_attr(all(target_os = "linux", any(target_env = "gnu", target_env = "uclibc")), repr(u32))]
    #[cfg_attr(any(
            bsd,
            target_os = "android",
            target_os = "aix",
            all(target_os = "linux", not(any(target_env = "gnu", target_env = "uclibc")))
        ), repr(i32))]
```

Reply:
Yeah unfortunately TryFrom is not implemented in this case... And I didn't actually notice that it's only u32 on Linux GNU/ulibc. Either way I think the way it is written now should be fine since i32 and u32 should be equivalent in memory. I also don't think `num_enum` would work since it is a proc macro and the enum is defined inside `nix`, inside of their `libc_enum` macro. Is there anything else we could do or just leave it as is?

Reply:
> Is there anything else we could do or just leave it as is?

Leaving it as-is looks good